hogben/models/samples.py

```diff
@@ -55,29 +71,44 @@ def __vary_structure(structure, bound_size=0.2):
             list: varying parameters of sample.
 
         """
-        params = []
-
-        # Vary the SLD and thickness of each component (layer).
-        for component in structure[1:-1]:
-            sld = component.sld.real
-            sld_bounds = (
-                sld.value * (1 - bound_size),
-                sld.value * (1 + bound_size),
-            )
-            sld.setp(vary=True, bounds=sld_bounds)
-            params.append(sld)
-
-            thick = component.thick
-            thick_bounds = (
-                thick.value * (1 - bound_size),
-                thick.value * (1 + bound_size),
-            )
-            thick.setp(vary=True, bounds=thick_bounds)
-            params.append(thick)
-
-        return params
-
-    def angle_info(self, angle_times, contrasts=None):
+        for structure in self.structures:
+            params = []
+            # The structure was defined in refnx.
+            if isinstance(structure, refnx.reflect.Structure):
+                # Vary the SLD and thickness of each component (layer).
+                for component in structure[1:-1]:
+                    sld = component.sld.real
+                    sld_bounds = (
+                        sld.value * (1 - bound_size),
+                        sld.value * (1 + bound_size),
```

Comment:
I think we should check if parameters already have bounds defined before assigning them

Reply:
Probably makes sense. I'm not 100% sure why we need `vary_structure` so it kinda depends on what the use-case is. But I think in principle it makes sense not to override previous bounds.

I kept it there not to break old workflows and implementations/notebooks. I think at the moment I mainly use it in unit tests...

Reply:
So I just looked this up, and by default (when not assigned any bounds) it sets the bounds to positive and negative infinity. I changed this somewhat to only use these newly redefined bounds when the bound is (+/-) infinite for each respective bound.

```
                if sld.bounds.lb == -float("inf"):
                    sld.bounds.lb = sld.value * (1 - bound_size)
                if sld.bounds.ub == float("inf"):
                    sld.bounds.ub = sld.value * (1 + bound_size)

                sld_bounds = (
                    sld.bounds.lb,
                    sld.bounds.ub,
                )
```

I could do the `if` conditional on a single line for each one, but that become pretty ugly to read.